go/plugins/ollama/ollama.go

```diff
@@ -0,0 +1,340 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ollama
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/firebase/genkit/go/ai"
+)
+
+const provider = "ollama"
+
+var roleMapping = map[ai.Role]string{
+	ai.RoleUser:   "user",
+	ai.RoleModel:  "assistant",
+	ai.RoleSystem: "system",
+}
+
+func defineModel(model ModelDefinition, serverAddress string) {
+	meta := &ai.ModelMetadata{
+		Label: "Ollama - " + model.Name,
+		Supports: ai.ModelCapabilities{
+			Multiturn:  model.Type == "chat",
+			SystemRole: model.Type != "chat",
+		},
+	}
+	g := &generator{model: model, serverAddress: serverAddress}
+	ai.DefineModel(provider, model.Name, meta, g.generate)
+}
+
+// Model returns the [ai.ModelAction] with the given name.
+// It returns nil if the model was not configured.
+func Model(name string) *ai.ModelAction {
+	return ai.LookupModel(provider, name)
+}
+
+// ModelDefinition represents a model with its name and type.
+type ModelDefinition struct {
+	Name string
+	Type string
+}
+
+// Config provides configuration options for the Init function.
+type Config struct {
+	// Server Address of oLLama.
+	ServerAddress string
+	// Generative models to provide.
+	Models []ModelDefinition
+}
+
+// Init registers all the actions in this package with ai.
+func Init(ctx context.Context, cfg Config) error {
+	for _, model := range cfg.Models {
+		defineModel(model, cfg.ServerAddress)
+	}
+	return nil
+}
+
+type generator struct {
+	model         ModelDefinition
+	serverAddress string
+}
+
+type ollamaMessage struct {
+	Role    string // json:"role"
+	Content string // json:"content"
+}
+
+/*
+TODO: Support optional, advanced parameters:
+format: the format to return a response in. Currently the only accepted value is json
+options: additional model parameters listed in the documentation for the Modelfile such as temperature
+system: system message to (overrides what is defined in the Modelfile)
+template: the prompt template to use (overrides what is defined in the Modelfile)
+context: the context parameter returned from a previous request to /generate, this can be used to keep a short conversational memory
+stream: if false the response will be returned as a single response object, rather than a stream of objects
+raw: if true no formatting will be applied to the prompt. You may choose to use the raw parameter if you are specifying a full templated prompt in your request to the API
+keep_alive: controls how long the model will stay loaded into memory following the request (default: 5m)
+*/
+type ollamaRequest struct {
```

Comment:
It's not clear at first reading what the difference is between ollamaRequest and ollamaGenerateRequest (and similar for Response). Adding a comment somewhere about the difference would help.

Reply:
Good point, I renamed it to ollamaChatRequest as well.